File: reuse_java/src/main/java/com/kmmania/Thrower.java
package com.kmmania;

/**
 * The Thrower class implements the {@link Athlete} class by composition and represents a thrower athlete.
 *
 * @author Dac Linh Giao NGUYEN
 * @version 0.1
 * @see Athlete
 */
public class Thrower {
    private final String firstName;
    private final String lastName;
    private final Athlete athlete;

    /**
     * Constructs a Thrower object with the specified first name, last name, and specialty.
     *
     * @param firstName  the first name of the thrower
     * @param lastName   the last name of the thrower
     * @param athlete    the athlete
     */
    public Thrower(String firstName, String lastName, Athlete athlete) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.athlete = athlete;
    }

    public String getFirstName() {
        return firstName;
    }

    public String getLastName() {
        return lastName;
    }

    /**
     * Displays the thrower's information including name and specialty.
     */
    public void display() {
        System.out.printf("My name is %s %s. %s\n", firstName, lastName, athlete.mySpeciality());
    }
}

File: reuse_python/src/thrower.py
"""
Athlete Reusability Project

This project demonstrates different ways to reuse classes in Python:
- Inheritance
- Composition
- Composition through interface implementation

Classes:
    Athlete: Represents an athlete with a specific specialty.
    Sprinter: Represents a sprinter, inheriting from Athlete.
    Thrower: Represents a thrower, utilizing Athlete class through composition.
    AthleteInterface: Interface defining the method for athlete specialty.
    AthleteImpl: Class implementing the AthleteInterface.
    DistanceRunner: Represents a distance runner, using Athlete interface through composition.
    Jumper: Represents a jumper, using Athlete interface through composition.

Usage:
    Import the required classes and create instances of different types of athletes.
    Each athlete instance can display information about themselves using the `display()` method.


Author: Dac Linh Giao NGUYEN
Date: 2024

"""
from src.athlete import Athlete


class Thrower:
    """Class representing a thrower.

    Attributes:
        first_name (str): The first name of the thrower.
        last_name (str): The last name of the thrower.
        athlete (Athlete): An Athlete object representing the thrower's specialty.

    Methods:
        display(): Displays information about the thrower, including their full name and specialty.
    """

    def __init__(self, first_name, last_name, athlete):
        """
        Initializes a Thrower object with a given first name, last name, and athlete.

        Args:
            first_name (str): The first name of the thrower.
            last_name (str): The last name of the thrower.
            athlete (Athlete): An Athlete object representing the thrower's specialty.
        """
        self.first_name = first_name
        self.last_name = last_name
        self.athlete = athlete

    def display(self):
        """
        Displays information about the thrower.

        Returns:
            None
        """
        print(f"My name is {self.first_name} {self.last_name}. {self.athlete.my_speciality()}")
